internal/ethapi/api.go

```diff
@@ -1652,50 +1622,48 @@ func (s *TransactionAPI) GetTransactionCount(ctx context.Context, address common
 // GetTransactionByHash returns the transaction for the given hash
 func (s *TransactionAPI) GetTransactionByHash(ctx context.Context, hash common.Hash) (*RPCTransaction, error) {
 	// Try to return an already finalized transaction
-	tx, blockHash, blockNumber, index, err := s.b.GetTransaction(ctx, hash)
-	if err != nil {
-		return nil, err
-	}
-	if tx != nil {
-		header, err := s.b.HeaderByHash(ctx, blockHash)
-		if err != nil {
-			return nil, err
+	found, tx, blockHash, blockNumber, index, err := s.b.GetTransaction(ctx, hash)
+	if !found {
+		// No finalized transaction, try to retrieve it from the pool
+		if tx := s.b.GetPoolTransaction(hash); tx != nil {
+			return NewRPCPendingTransaction(tx, s.b.CurrentHeader(), s.b.ChainConfig()), nil
+		}
+		if err == nil {
+			return nil, nil
 		}
-		return newRPCTransaction(tx, blockHash, blockNumber, header.Time, index, header.BaseFee, s.b.ChainConfig()), nil
+		return nil, NewTxIndexingError()
 	}
-	// No finalized transaction, try to retrieve it from the pool
-	if tx := s.b.GetPoolTransaction(hash); tx != nil {
-		return NewRPCPendingTransaction(tx, s.b.CurrentHeader(), s.b.ChainConfig()), nil
+	header, err := s.b.HeaderByHash(ctx, blockHash)
+	if err != nil {
+		return nil, err
 	}
-
-	// Transaction unknown, return as such
-	return nil, nil
+	return newRPCTransaction(tx, blockHash, blockNumber, header.Time, index, header.BaseFee, s.b.ChainConfig()), nil
 }
 
 // GetRawTransactionByHash returns the bytes of the transaction for the given hash.
 func (s *TransactionAPI) GetRawTransactionByHash(ctx context.Context, hash common.Hash) (hexutil.Bytes, error) {
 	// Retrieve a finalized transaction, or a pooled otherwise
-	tx, _, _, _, err := s.b.GetTransaction(ctx, hash)
-	if err != nil {
-		return nil, err
-	}
-	if tx == nil {
-		if tx = s.b.GetPoolTransaction(hash); tx == nil {
-			// Transaction not found anywhere, abort
+	found, tx, _, _, _, err := s.b.GetTransaction(ctx, hash)
+	if !found {
+		if tx = s.b.GetPoolTransaction(hash); tx != nil {
+			return tx.MarshalBinary()
+		}
+		if err == nil {
 			return nil, nil
 		}
+		return nil, NewTxIndexingError()
 	}
-	// Serialize to RLP and return
 	return tx.MarshalBinary()
 }
 
 // GetTransactionReceipt returns the transaction receipt for the given transaction hash.
 func (s *TransactionAPI) GetTransactionReceipt(ctx context.Context, hash common.Hash) (map[string]interface{}, error) {
-	tx, blockHash, blockNumber, index, err := s.b.GetTransaction(ctx, hash)
-	if tx == nil || err != nil {
```

Comment:
Could geth team reconsider this change?

It breaks hardhat.

```
An unexpected error occurred:

ProviderError: transaction indexing is in progress
    at HttpProvider.request (/workspace/node_modules/hardhat/src/internal/core/providers/http.ts:96:21)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async EIP1193JsonRpcClient.getTransactionReceipt (/workspace/node_modules/@nomicfoundation/ignition-core/src/internal/execution/jsonrpc-client.ts:569:22)
    at async Promise.all (index 1)
    at async monitorOnchainInteraction (/workspace/node_modules/@nomicfoundation/ignition-core/src/internal/execution/future-processor/handlers/monitor-onchain-interaction.ts:110:28)
    at async FutureProcessor.processFuture (/workspace/node_modules/@nomicfoundation/ignition-core/src/internal/execution/future-processor/future-processor.ts:116:9)
    at async ExecutionEngine._executeBatch (/workspace/node_modules/@nomicfoundation/ignition-core/src/internal/execution/execution-engine.ts:153:30)
    at async ExecutionEngine.executeModule (/workspace/node_modules/@nomicfoundation/ignition-core/src/internal/execution/execution-engine.ts:114:25)
    at async Deployer.deploy (/workspace/node_modules/@nomicfoundation/ignition-core/src/internal/deployer.ts:194:25)
    at async SimpleTaskDefinition.action (/workspace/node_modules/@nomicfoundation/hardhat-ignition/src/index.ts:302:24)
```

Reply:
I'm also unhappy with this, and there have been quite a few reports from across the ecosystem where this PR caused subtle issues. However, we ultimately decided to go ahead with the PR because we also changed eth_syncing to take indexing into account. Not sure if that works for hardhat...

Reply:
A better way is to return nil as before.

Reply:
We must to distinguish two different scenarios here:

- The requested transaction/receipt is existent but not indexed yet
- The requested transaction/receipt is unknown to node

Therefore, we explicitly return an error if the background indexing is still in progress
and the requested transaction is not indexed yet.

Reply:
I don't think so.

> We must

It's a breaking change, geth is a dominant ethereum client, every breaking change needs to consider backward compatibility.

> The requested transaction/receipt is existent but not indexed yet

The code just checks if the index is running, it means geth can't know if the transaction is indexing or not, what did I miss?

```
	tx, blockHash, blockNumber, txIndex := rawdb.ReadTransaction(bc.db, hash)
	if tx == nil {
		progress, err := bc.TxIndexProgress()
		if err != nil {
			return nil, nil, nil
		}
		// The transaction indexing is not finished yet, returning an
		// error to explicitly indicate it.
		if !progress.Done() {
			return nil, nil, errors.New("transaction indexing still in progress")
		}
		// The transaction is already indexed, the transaction is either
		// not existent or not in the range of index, returning null.
		return nil, nil, nil
	}
```

Reply:
In both scenarios mentioned, no transaction index will be found. The error will simply *state, Please try the RPC later once background indexing is fully complete*.

> Since Geth is a dominant Ethereum client, every breaking change must consider backward compatibility. 

That's why we've modified the eth_syncing endpoint. Transaction indexing is also taken into account and will only return "synced" once all indexing is complete.